File: 2021/day-04/2021-04-1.py
#!/usr/bin/env python3
import sys
sys.path.append("../../libs/python")
import n0s1_aoc as n0s1


#   0   1   2   3   4
#   5   6   7   8   9
#  10  11  12  13  14
#  15  16  17  18  19
#  20  21  22  23  24

def solved(board):
    for i in range(0,5):
        row_valid = True
        column_valid = True
        for j in range((i*5),5+(i*5)):
            if board[j] != "X":
                row_valid = False
        for j in range(0+i,21+i, 5):
            if board[j] != "X":
                column_valid = False
        if row_valid or column_valid:
            return True
    return False

def call_numbers(input, boards):
    for i in input:
        for b, board in enumerate(boards):
            for t, tile in enumerate(board):
                if tile == i:
                    boards[b][t] = "X"
            if solved(boards[b]):
                return [b, i]

def bingo_time(data):
    sdata = data.split("\n\n")
    input = sdata[0].split(",")
    boards = [board.replace("\n", " ").replace("  ", " ").split(" ") for board in sdata[1:]]
    winner = call_numbers(input, boards)
    sum = 0
    for i in boards[winner[0]]:
        if i != "X" and i != "":
            sum += int(i)

    return sum * int(winner[1])

print(bingo_time(n0s1.file_to_string("sample-2021-04.txt")))
print(bingo_time(n0s1.file_to_string("input-2021-04.txt")))

File: 2021/day-04/2021-04-1.ex
defmodule Day4 do
  import N0S1_AOC
  use Bitwise
  @win_conds [31, 992, 31744, 1015808, 32505856, 1082401,
              2164802, 4329604, 8659208, 17318416]

  def bingo_time(fname) do
    input = read_file(fname) |> String.split("\n\n")
    calls = hd(input) |> String.split(",")
    boards = tl(input) |> Enum.map(fn board ->
      String.replace(board, "  ", " ") |> String.split()
    end)

    {last_call, winning_state, winning_board} = call_numbers(calls, boards)

    # Apply bitmask to remove called numbers
    not_called = Enum.reverse(int_to_bitlist(winning_state, 25))
    |> Enum.zip(winning_board)
    |> Enum.flat_map(fn {bit, board_cell} ->
      if bit == 0 do [String.to_integer(board_cell)] else [] end
    end)

    Enum.sum(not_called) * String.to_integer(last_call)
  end

  def call_numbers(calls, boards) do
    states = List.duplicate(0, length(boards))
    |> Enum.with_index
    |> Enum.zip(boards)
    call_numbers(calls, states, nil)
  end

  def call_numbers([num | t], boards, winner) do
    try do
      new_boards = Enum.map(boards, fn {{state, idx}, board} ->
        match = Enum.find_index(board, fn cell -> cell == num end)
        if match == nil do
          {{state, idx}, board}
        else
          new_state = state + 2 ** match
          if Enum.any?(@win_conds, fn wc -> band(new_state, wc) == wc end) do
            throw({:winner, {num, new_state, board}})
          else
            {{new_state, idx}, board}
          end
        end
      end)
      call_numbers(t, new_boards, winner)
    catch
      {:winner, w} -> w
    end
  end

  def main do
    IO.puts bingo_time("sample-2021-04.txt")
    IO.puts bingo_time("input-2021-04.txt")
  end

end
